// @modern-dev/jsbn/lib/cjs/index.js
"use strict";
/*!
 * JavaScript BigInteger
 * https://github.com/modern-dev/jsbn
 *
 * Copyright (c) 2019 Bohdan Shtepan
 * Licensed under a BSD license.
 *
 * Original copyright
 * Copyright (c) 2003-2005 Tom Wu
 * All Rights Reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const secure_random_1 = require("./secure-random");
exports.SecureRandom = secure_random_1.default;
exports.Arcfour = secure_random_1.Arcfour;
exports.rng_seed_time = secure_random_1.rng_seed_time;
const algorithms_1 = require("./algorithms");
exports.Classic = algorithms_1.Classic;
exports.Barrett = algorithms_1.Barrett;
exports.Montgomery = algorithms_1.Montgomery;
exports.NullExp = algorithms_1.NullExp;
const big_integer_1 = require("./big-integer");
exports.BigInteger = big_integer_1.default;
exports.default = {
    Arcfour: secure_random_1.Arcfour,
    SecureRandom: secure_random_1.default,
    rng_seed_time: secure_random_1.rng_seed_time,
    Barrett: algorithms_1.Barrett,
    Classic: algorithms_1.Classic,
    Montgomery: algorithms_1.Montgomery,
    NullExp: algorithms_1.NullExp,
    BigInteger: big_integer_1.default
};
